pkg/policy/directory/watcher.go

```diff
@@ -0,0 +1,207 @@
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+
+	ipcacheTypes "github.com/cilium/cilium/pkg/ipcache/types"
+	k8sConst "github.com/cilium/cilium/pkg/k8s/apis/cilium.io"
+	k8sCiliumUtils "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/utils"
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+type policyWatcher struct {
+	log           logrus.FieldLogger
+	config        *option.DaemonConfig
+	policyManager PolicyManager
+	// maps cnp file name to cnp object. this is required to retrieve data during delete.
+	fileNameToCnpCache map[string]*cilium_v2.CiliumNetworkPolicy
+}
+
+func (p *policyWatcher) translateToCNPObject(file string) (*cilium_v2.CiliumNetworkPolicy, error) {
+	data, err := os.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+
+	// yaml to json conversion
+	jsonData, err := yaml.YAMLToJSON([]byte(data))
+	if err != nil {
+		return nil, err
+	}
+
+	// translate json to cnp object
+	cnp := &cilium_v2.CiliumNetworkPolicy{}
+	err = json.Unmarshal(jsonData, cnp)
+	if err != nil {
+		return nil, err
+	}
+
+	return cnp, nil
+}
+
+func getLabels(filePath string, cnp *cilium_v2.CiliumNetworkPolicy) labels.LabelArray {
+	name := cnp.ObjectMeta.Name
+	ns := cnp.ObjectMeta.Namespace
+
+	labelsArr := labels.LabelArray{
+		labels.NewLabel("name", name, labels.LabelSourceDirectory),
+		labels.NewLabel("path", filePath, labels.LabelSourceDirectory),
+	}
+
+	// For clusterwide policy namespace will be empty.
+	if ns != "" {
+		nsLabel := labels.NewLabel(k8sConst.PolicyLabelNamespace, ns, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, nsLabel)
+		srcLabel := labels.NewLabel("policy-derived-from", k8sCiliumUtils.ResourceTypeCiliumNetworkPolicy, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, srcLabel)
+	} else {
+		srcLabel := labels.NewLabel("policy-derived-from", k8sCiliumUtils.ResourceTypeCiliumClusterwideNetworkPolicy, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, srcLabel)
+	}
+
+	return labelsArr
+}
+
+// read cilium network policy yaml file and convert to policy object and
+// add rules to policy engine.
+func (p *policyWatcher) addToPolicyEngine(cnpFilePath string) error {
+	// read from file and convert to cnp object
+	cnp, err := p.translateToCNPObject(cnpFilePath)
+	if err != nil {
+		return err
+	}
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindCNP,
+		cnp.ObjectMeta.Namespace,
+		cnp.ObjectMeta.Name,
+	)
+
+	// convert to rules
+	rules, err := cnp.Parse()
+	if err != nil {
+		return err
+	}
+
+	//update labels
+	for _, r := range rules {
+		r.Labels = getLabels(cnpFilePath, cnp)
+	}
+
+	// add to policy engine
+	_, err = p.policyManager.PolicyAdd(rules, &policy.AddOptions{
+		ReplaceWithLabels: getLabels(cnpFilePath, cnp),
```

Comment:
We actually maybe want to use `ReplaceByResource: true` here instead (assuming we use the filename as the resourceID as suggested above) and leave `ReplaceWithLabels` empty. This way, the policy lifetime is tied to the file itself (which is I think what we want).

`ReplaceWithLabels` is mainly (only?) used when importing a policy from API, where we don't have a unique resource handle for the API request. But when importing policies from CNP, we also use `ReplaceByResource`